### ISSUE TITLE
Threaded conversations

### ISSUE THREAD

**Issue by ktoso** [gitterHQ/gitter#741]
We are trying to switch from Flowdock to Gitter in order to open up our internal communication to the wider developer community (we as in @akka), however one of the killer features we loved in Flowdock is not available in gitter: Threaded discussions.

They are very important to not be completely confused when catching up on a discussion which one did not track in real time, and also allow to ignore a discussion I don't care about (at this moment). For example me and another dev discuss feature X in one "thread", so we want to see only responses in that "thread" (and pop back out to the "see everything" easily), while others discuss features Y and Z, all in parallel. Without threads the amount of noise makes it hard to track these multiple independent conversations.

Are you guys thinking of implementing something that would address this? 
It is even more visible the more popular an open source project is, as many people chime in and there's usually at least a few discussions in parallel.

**Comment by mydigitalself** [comment]
Hi, we absolutely are and hope to have this done by the end of the summer.

**Comment by mydigitalself** [comment]
PS, we often handle this internally by creating channels per features too.

**Comment by ktoso** [comment]
> Hi, we absolutely are and hope to have this done by the end of the summer.

That's wonderful to hear. I'll be cheering on you guys then. Happy to help if you're at a point help could be needed.

> PS, we often handle this internally by creating channels per features too.

Right, that works as semi-solution. Looking forward to the full-blown thing, it's a killer feature for us :-)

**Comment by folex** [comment]
`+1` for threaded discussions, it will help a lot in big channels where multiple topics are discussed simultaneously.

**Comment by tdhd** [comment]
also :+1: from me for a great feature.

**Comment by jducoeur** [comment]
Yeah, between the akka and scala-js communities I spend a *lot* of time in Gitter, and this is clearly the feature that would most improve the experience.  I think we have multiple discussions going simultaneously more often than not...

**Comment by mydigitalself** [comment]
We've been thinking about this a lot and it's very difficult to do. Flowdock's approach is the closest we've seen, and it still has issues if people don't explicitly reply to the messages or if people accidentally leave themselves in reply mode, then you're left with a pretty messy situation.

We're still trying to figure out it, any suggestions would be welcome :)

**Comment by jducoeur** [comment]
I assume the problem is more UI/UX, rather than technical?

**Comment by mydigitalself** [comment]
Absolutely.

**Comment by jducoeur** [comment]
Hmm.  I haven't used Flowdock, so I don't know what insights they might have had, but this is [one of my long-time professional interests](https://artofconv.wordpress.com/), so I'm happy to help brainstorm ideas.  I suspect you've thought of most of this, but a few observations off the top of my head:

* In an IM-ish environment like this, threads are typically *discovered*, rather than intentionally started.  That is, it isn't that I send a message saying "I'm starting a thread"; moreover, it isn't as if the first person to *reply* to me thinks "I'm replying to a thread".  It's that five replies later, I realize "Oh, that's a thread".  So ideally, I want to way to *quickly* mark the preceding messages and pull them out into a thread.  Indeed, it's often someone who *isn't* participating in the thread who is motivated to want that, to pull out some comments into their own thread as a way of saying, "Y'all take this out of the main conversation, please".

* *Ideally*, you want the ability to define sub-threads, since threads will develop tangents.  I suspect that's not critical for your use case -- simply creating a new parallel thread might suffice -- but you might want to be think in terms of some sort of tree structure under the hood, with threads emerging from other threads at particular points in the stream.  (This is how I handle conversation in my system, Querki, although the threading there is more explicit instead of discovered.)

* Threads really want to be not just visually distinct, but visually *separate* in order to be comprehensible.  That gets into issues of screen layout which aren't trivial -- again, thinking of the thread as being rooted at a point in the overall stream may be the best way to do it, but there are a bunch of alternatives.

* More importantly, being able to find new messages in threads quickly, and navigate to them, becomes a bit tricky -- it's no longer a one-dimensional "39 new messages up there".  Google Wave did pretty well on this score, IMO: spacebar navigated from blip to blip, starting with the oldest unread comment, then through that thread, then on to the next thread.  I'm not sure there's a better approach -- that seems to be the logical way to expect to keep up with a multi-threaded conversation.  (Wave tied itself in knots with its comments-within-comments-within-comments capability, but had a lot of good ideas that are worth exploiting.)

* Once we have threading, I immediately want the ability to be able to tune out a thread -- to mute it so that the system stops notifying me about new content in it, and probably collapses it visually.  You might as well plan for that upfront.

That's what comes to mind in the first ten minutes, anyway.  Happy to keep kibitzing if you want the outside viewpoint -- it's an interesting problem...

**Comment by pitrou** [comment]
+1 for this. Copying Flowdock's UI would be enough for us.

**Comment by ebekker** [comment]
Guess this feature didn't make it in time for end of summer, any ETA?

**Comment by ebekker** [comment]
So the problem with basecamp is that it's inherently a closed system, where someone needs to add others to it.  I'm looking for an oss-friendly way that anyone can access the topics, even search and read anonymously, and participate if they want to.  Again, wtih Gitter this is nice because it integrates with GitHub's for a user's identity and it's completely open for anyone to participate.

**Comment by IRus** [comment]
+1 will be happy to have threaded discussions in chat. This is must have feature for communities chats.

**Comment by egeland** [comment]
+1

Just blatantly borrow FlowDock's way:
Clicking 'reply' on a line either starts that as a new conversation or joins the conversation that line is part of. Set a thread colour (just cycle through a bunch of decent-looking colours) at pseudo-random.

**Comment by abhisekp** [comment]
:+1: This would be a really nice feature to have.

**Comment by TheJaredWilcurt** [comment]
I think it would be better, for realtime chat, to be able to "Reply-To" a specific message, and then your message simply has a "View Parent" option that would expand and show an embedded version of the message you are replying to if it is off screen, otherwise it would just highlight it if it's on screen. If it's off screen there would be an option to click on it and take you to see the actual message so you could view it's parent in the same method. This is a better solution as not everyone will "Reply To" everyone else, so it would take you to the message 4 messages up the chain and you could see the context there. You could even detect the most recent message from the person being @ referenced and then set that as the thread parent.
This keeps the feel of a chatroom, but adds in the benefits of tracking a threaded conversation

**Comment by jducoeur** [comment]
Probably is, requiring explicit Reply-To every time won't work when you have two or more realtime conversations happening at once, which happens quite often in Gitter -- people just *won't* remember to press the Reply-To button more than occasionally, and it all turns into a hash.  (Which in fact is what happens now -- a few people think to @-refer upthread, but most don't.)  You need some way to pull the conversations apart visually at the time you're trying to deal with them, in order to keep them straight.

Mind, I love Gitter, and spend hours a day in it.  But *man*, those interleaved conversations get hard to follow sometimes...

**Comment by dhirschfeld** [comment]
+1 - FlowDocks solution works well but it would be nice to be able to move a message to a thread after you send it in the case that you forgot to reply in the thread before sending

**Comment by mydigitalself** [comment]
From the Twitter conversation...

https://twitter.com/Sander_Spies/status/763310662600851457

We're working on something called Topics right now, which is more threaded type of conversations - but they sit alongside a Community rather than being in chat. So basically an area in your community where you can have threaded FAQs, perhaps documentation, Q&A etc... 

Our current thinking on threading chat is you can then take that chat conversations and convert it into a topic for longevity - although this will only come a little later.

**Comment by jmcmurry** [comment]
+1
